clarity/src/vm/tests/traits.rs

```diff
@@ -1281,3 +1286,345 @@ fn test_return_trait_with_contract_of(owned_env: &mut OwnedEnvironment) {
         );
     }
 }
+
+fn test_pass_trait_to_subtrait(owned_env: &mut OwnedEnvironment) {
+    let dispatching_contract = "(define-trait trait-1 (
+            (get-1 (uint) (response uint uint))
+        ))
+        (define-trait trait-12 (
+            (get-1 (uint) (response uint uint))
+            (get-2 (uint) (response uint uint))
+        ))
+        (define-public (wrapped-get-1 (contract <trait-12>))
+            (internal-get-1 contract))
+        (define-public (internal-get-1 (contract <trait-1>))
+            (contract-call? contract get-1 u1))";
+    let target_contract = "(define-public (get-1 (a uint)) (ok a))
+        (define-public (get-2 (a uint)) (ok a))";
+
+    let p1 = execute("'SZ2J6ZY48GV1EZ5V2V5RB9MP66SW86PYKKQ9H6DPR");
+    let mut placeholder_context = ContractContext::new(
+        QualifiedContractIdentifier::transient(),
+        ClarityVersion::Clarity2,
+    );
+
+    {
+        let mut env = owned_env.get_exec_environment(None, None, &mut placeholder_context);
+        env.initialize_contract(
+            QualifiedContractIdentifier::local("dispatching-contract").unwrap(),
+            dispatching_contract,
+        )
+        .unwrap();
+
+        env.initialize_contract(
+            QualifiedContractIdentifier::local("target-contract").unwrap(),
+            target_contract,
+        )
+        .unwrap();
+    }
+
+    {
+        let target_contract = Value::from(PrincipalData::Contract(
+            QualifiedContractIdentifier::local("target-contract").unwrap(),
+        ));
+        let mut env = owned_env.get_exec_environment(
+            Some(p1.clone().expect_principal()),
+            None,
+            &mut placeholder_context,
+        );
+        assert_eq!(
+            env.execute_contract(
+                &QualifiedContractIdentifier::local("dispatching-contract").unwrap(),
+                "wrapped-get-1",
+                &symbols_from_values(vec![target_contract]),
+                false
+            )
+            .unwrap(),
+            Value::okay(Value::UInt(1)).unwrap()
+        );
+    }
+}
+
+fn test_embedded_trait(owned_env: &mut OwnedEnvironment) {
```

Comment:
Related to embedded contracts, is it possible to create a contract that implements a trait it defines, and then calls a function in itself that takes itself as an argument, and in doing so, create a function call loop?  Something like this:

```clarity
(define-trait trait-1 (
   (echo (uint) (response uint uint))
))
(define-public (echo (x uint))
   ;; $this-contract-address can be filled in prior to deployment
   (ok (do-echo $this-contract-address x))
)
(define-public (do-echo (echoer <trait-1>) (arg uint))
  (contract-call? echoer echo arg)
)
```

What happens if someone calls `echo`?

Reply:
I don't think it is possible to fill in `$this-contract-address`, is it? Wouldn't this always give you a `NoSuchContract` error? If you tried to use `(as-contract tx-sender)`, that gets a `Principal` type, not a `Callable`, so you will get an error.

Reply:
Yup, makes sense.  Thanks!